### ISSUE TITLE
[rllib] " can't pickle _LazyLoader objects"

### ISSUE THREAD

**Issue by zzchuman** [ray-project/ray#13051]
Hello, I custom the centarlized critic model via PPO. And my version is: tf 2.0.0, ray 1.0.0, python 3.7.

But, it shows " can't pickle _LazyLoader objects"

![image](https://user-images.githubusercontent.com/38434094/102953259-d70f6800-450b-11eb-9efa-5240e57a8831.png)

I don't know what's wrong？

import argparse
from ray.rllib.models.modelv2 import ModelV2
from ray.rllib.agents.ppo.ppo_tf_policy import PPOTFPolicy, KLCoeffMixin, ppo_surrogate_loss as tf_loss
from ray.rllib.evaluation.postprocessing import compute_advantages, Postprocessing
from ray.rllib.models import ModelCatalog
from ray.rllib.policy.sample_batch import SampleBatch
from ray.rllib.policy.tf_policy import LearningRateSchedule, EntropyCoeffSchedule
from ray.rllib.utils.tf_ops import explained_variance, make_tf_callable
from ray.rllib.models.tf.tf_modelv2 import TFModelV2
from ray.rllib.models.tf.fcnet import FullyConnectedNetwork
from ray.rllib.utils.annotations import override
from ray.rllib.policy.torch_policy import LearningRateSchedule as TorchLR, \
    EntropyCoeffSchedule as TorchEntropyCoeffSchedule
import ray
from ray import tune
from ray.rllib.agents.ppo.ppo import PPOTrainer
from ray.tune.registry import register_env
from ray.tune.logger import pretty_print
from ray.tune import grid_search
from gym import spaces
from sumo_rl.environment.env02 import SumoEnvironment
import traci
import numpy as np
from ray.rllib.utils.framework import try_import_torch
from ray.rllib.utils.framework import try_import_tf
torch, nn = try_import_torch()
torch.manual_seed(42)
np.random.seed(65)
tf1, tf, tfv = try_import_tf()

class CentralizedCriticModel(TFModelV2):
    """Multi-agent model that implements a centralized value function."""
    def __init__(self, obs_space, action_space, num_outputs, model_config, name):
        super(CentralizedCriticModel, self).__init__(obs_space, action_space, num_outputs, model_config, name)

        # Base of the model
        self.model = FullyConnectedNetwork(obs_space, action_space, num_outputs, model_config, name)
        self.register_variables(self.model.variables())

        n_agents = 9  
        obs = 20
        act = 4
        opp_obs_accum = obs * (n_agents - 1)
        opp_acts_accum = act * (n_agents - 1)

        # Central VF maps (obs, opp_obs, opp_act) -> vf_pred
        obs = tf.keras.layers.Input(shape=(obs,), name="obs")
        opp_obs = tf.keras.layers.Input(shape=(opp_obs_accum,), name="opp_obs")
        opp_act = tf.keras.layers.Input(shape=(opp_acts_accum,), name="opp_act")
        concat_obs = tf.keras.layers.Concatenate(axis=1)([obs, opp_obs, opp_act])
        central_vf_dense = tf.keras.layers.Dense(16, activation=tf.nn.tanh, name="c_vf_dense")(concat_obs)
        central_vf_out = tf.keras.layers.Dense(1, activation=None, name="c_vf_out")(central_vf_dense)

        self.central_vf = tf.keras.Model(inputs=[obs, opp_obs, opp_act],outputs=central_vf_out)
        self.register_variables(self.central_vf.variables)

    @override(ModelV2)
    def forward(self, input_dict, state, seq_lens):
        return self.model.forward(input_dict, state, seq_lens)

    def central_value_function(self, obs, opponent_obs, opponent_actions):
        return tf.reshape(self.central_vf([obs, opponent_obs, tf.one_hot(opponent_actions, 4)]),[-1])  # ---> changed the depth of one_hot encoding to 5 (5 actions)

    @override(ModelV2)
    def value_function(self):
        return self.model.value_function()  # not used

OPPONENT_OBS = "opponent_obs"
OPPONENT_ACTION = "opponent_action"

class CentralizedValueMixin:
    """Add method to evaluate the central value function from the model."""

    def __init__(self):
        if self.config["framework"] != "torch":
            self.compute_central_vf = make_tf_callable(self.get_session())(
                self.model.central_value_function)

# Grabs the opponent obs/act and includes it in the experience train_batch, and computes GAE using the central vf predictions.
def centralized_critic_postprocessing(policy,
                                      sample_batch,
                                      other_agent_batches=None,
                                      episode=None):
    if policy.loss_initialized():
        assert other_agent_batches is not None
        # [(_, opponent_batch)] = list(other_agent_batches.values())

        # ---> opponent batch now consists of 8 SampleBatches, so I concatenate them
        concat_opponent_batch = SampleBatch.concat_samples([opponent_n_batch
                                                            for _, opponent_n_batch in other_agent_batches.values()])

        opponent_batch = concat_opponent_batch

        # also record the opponent obs and actions in the trajectory
        sample_batch[OPPONENT_OBS] = opponent_batch[SampleBatch.CUR_OBS]
        sample_batch[OPPONENT_ACTION] = opponent_batch[SampleBatch.ACTIONS]

        sample_batch[SampleBatch.VF_PREDS] = policy.compute_central_vf(
            sample_batch[SampleBatch.CUR_OBS], sample_batch[OPPONENT_OBS],
            sample_batch[OPPONENT_ACTION])
    else:
        # Policy hasn't been initialized yet, use zeros.
        sample_batch[OPPONENT_OBS] = np.zeros_like(
            sample_batch[SampleBatch.CUR_OBS])
        sample_batch[OPPONENT_ACTION] = np.zeros_like(
            sample_batch[SampleBatch.ACTIONS])
        sample_batch[SampleBatch.VF_PREDS] = np.zeros_like(
            sample_batch[SampleBatch.REWARDS], dtype=np.float32)

    completed = sample_batch["dones"][-1]
    if completed:
        last_r = 0.0
    else:
        last_r = sample_batch[SampleBatch.VF_PREDS][-1]

    train_batch = compute_advantages(
        sample_batch,
        last_r,
        policy.config["gamma"],
        policy.config["lambda"],
        use_gae=policy.config["use_gae"])
    return train_batch

# Copied from PPO but optimizing the central value function.
def loss_with_central_critic(policy, model, dist_class, train_batch):
    CentralizedValueMixin.__init__(policy)
    func = tf_loss

    vf_saved = model.value_function

    model.value_function = lambda: policy.model.central_value_function(
        train_batch[SampleBatch.CUR_OBS], train_batch[OPPONENT_OBS],
        train_batch[OPPONENT_ACTION])

    policy._central_value_out = model.value_function()
    loss = func(policy, model, dist_class, train_batch)

    model.value_function = vf_saved

    return loss

def setup_torch_mixins(policy, obs_space, action_space, config):
    # Copied from PPOTorchPolicy  (w/o ValueNetworkMixin).
    TorchKLCoeffMixin.__init__(policy, config)
    TorchEntropyCoeffSchedule.__init__(policy, config["entropy_coeff"],
                                       config["entropy_coeff_schedule"])
    TorchLR.__init__(policy, config["lr"], config["lr_schedule"])

def setup_tf_mixins(policy, obs_space, action_space, config):
    # Copied from PPOTFPolicy (w/o ValueNetworkMixin).
    KLCoeffMixin.__init__(policy, config)
    EntropyCoeffSchedule.__init__(policy, config["entropy_coeff"],
                                  config["entropy_coeff_schedule"])
    LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])

def central_vf_stats(policy, train_batch, grads):
    # Report the explained variance of the central value function.
    return {
        "vf_explained_var": explained_variance(
            train_batch[Postprocessing.VALUE_TARGETS],
            policy._central_value_out),
    }

CCPPOTFPolicy = PPOTFPolicy.with_updates(
    name="CCPPOTFPolicy",
    postprocess_fn=centralized_critic_postprocessing,
    loss_fn=loss_with_central_critic,
    before_loss_init=setup_tf_mixins,
    grad_stats_fn=central_vf_stats,
    mixins=[
        LearningRateSchedule, EntropyCoeffSchedule, KLCoeffMixin,
        CentralizedValueMixin
    ])

def get_policy_class(config):
    return CCPPOTFPolicy

CCTrainer = PPOTrainer.with_updates(
    name="CCPPOTrainer",
    default_policy=CCPPOTFPolicy,
    get_policy_class=get_policy_class,
)

def policy_mapping(id):
    if id == '0':
        return '0'
    elif id == '1':
        return '1'
    elif id == '2':
        return '2'
    elif id == '3':
        return '3'
    elif id == '4':
        return '4'
    elif id == '5':
        return '5'
    elif id == '6':
        return '6'
    elif id == '7':
        return '7'
    else:
        return '8'

ModelCatalog.register_custom_model("cc_model", CentralizedCriticModel)

if __name__ == "__main__":
    ray.init(local_mode=True, include_dashboard=False)
    model_dir = 'outputs/3x3ppoicmgrid1/ppo-contexts'
    if not os.path.exists(model_dir):
        os.makedirs(model_dir)

    # Register the model and environment
    grouping = {
        'group_1': ['0', '1', '2', '3', '4', '5', '6', '7', '8']
    }
    parser = argparse.ArgumentParser()
    parser.add_argument("--framework", type=str, default=tf)
    args = parser.parse_args()

    register_env("3x3grid", lambda _: SumoEnvironment(net_file='nets/3x3grid/3x3Grid2lanes350.net.xml',
                                                      route_file='nets/3x3grid/3x3p318.rou.xml',
                                                      out_csv_name='outputs/3x3ppoicmgrid1/ppo-contexts',
                                                      use_gui=False,
                                                      num_seconds=18000,
                                                      time_to_load_vehicles=120,
                                                      max_depart_delay=0,
                                                      phases=[
                                                          traci.trafficlight.Phase(30, "GGGrrrrrGGGrrrrr"),
                                                          traci.trafficlight.Phase(3, "YYYrrrrrYYYrrrrr"),
                                                          traci.trafficlight.Phase(15, "rrrGrrrrrrrGrrrr"),
                                                          traci.trafficlight.Phase(3, "rrryrrrrrrryrrrr"),
                                                          traci.trafficlight.Phase(30, "rrrrGGGrrrrrGGGr"),
                                                          traci.trafficlight.Phase(3, "rrrrYYYrrrrrYYYr"),
                                                          traci.trafficlight.Phase(15, "rrrrrrrGrrrrrrrG"),
                                                          traci.trafficlight.Phase(3, "rrrrrrryrrrrrrry")
                                                      ]))

    config = {
        "env": "3x3grid",
        "batch_mode": "complete_episodes",
        "num_workers": 0,
        "lr": 0.0002,
        "seed": 42,
        "multiagent": {
            "policies": {
                '0': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {}),
                '1': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {}),
                '2': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {}),
                '3': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {}),
                '4': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {}),
                '5': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {}),
                '6': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {}),
                '7': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {}),
                '8': (spaces.Box(low=np.zeros(20), high=np.array(['inf'] * 20)), spaces.Discrete(4), {})
            },
            "policy_mapping_fn": policy_mapping  # Traffic lights are always controlled by this policy
        },
        "model": {
            "custom_model": "cc_model",
        },
        "framework": "tf",
    }

    stop = {
        "training_iteration": args.stop_iters,
        "timesteps_total": args.stop_timesteps,
        "episode_reward_mean": args.stop_reward,
    }

    results = tune.run(CCTrainer,name="TRAFFIC",
                       config=config,stop=stop,verbose=1,checkpoint_freq=10,
                       checkpoint_at_end=True,mode="max",metric="episode_reward_mean")

    print(f"Best checkpoint at: {results.best_checkpoint}")

**Comment by rkooo567** [comment]
Can you try using the latest master and see if you can troubleshoot with this tool? 

https://docs.ray.io/en/master/serialization.html#troubleshooting
